// MDB-React-Filter/dist/mdb-react-filter.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var reactTransitionGroup = require('react-transition-group');
var ReactResizeDetector = _interopDefault(require('react-resize-detector'));
var Lightbox = _interopDefault(require('react-image-lightbox'));
var uuid = _interopDefault(require('uuid'));

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _isNativeReflectConstruct() {
  if (typeof Reflect === "undefined" || !Reflect.construct) return false;
  if (Reflect.construct.sham) return false;
  if (typeof Proxy === "function") return true;

  try {
    Date.prototype.toString.call(Reflect.construct(Date, [], function () {}));
    return true;
  } catch (e) {
    return false;
  }
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _createSuper(Derived) {
  return function () {
    var Super = _getPrototypeOf(Derived),
        result;

    if (_isNativeReflectConstruct()) {
      var NewTarget = _getPrototypeOf(this).constructor;

      result = Reflect.construct(Super, arguments, NewTarget);
    } else {
      result = Super.apply(this, arguments);
    }

    return _possibleConstructorReturn(this, result);
  };
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

var ReactPropTypesSecret = 'SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED';

var ReactPropTypesSecret_1 = ReactPropTypesSecret;

function emptyFunction() {}
function emptyFunctionWithReset() {}
emptyFunctionWithReset.resetWarningCache = emptyFunction;

var factoryWithThrowingShims = function() {
  function shim(props, propName, componentName, location, propFullName, secret) {
    if (secret === ReactPropTypesSecret_1) {
      // It is still safe when called from React.
      return;
    }
    var err = new Error(
      'Calling PropTypes validators directly is not supported by the `prop-types` package. ' +
      'Use PropTypes.checkPropTypes() to call them. ' +
      'Read more at http://fb.me/use-check-prop-types'
    );
    err.name = 'Invariant Violation';
    throw err;
  }  shim.isRequired = shim;
  function getShim() {
    return shim;
  }  // Important!
  // Keep this list in sync with production version in `./factoryWithTypeCheckers.js`.
  var ReactPropTypes = {
    array: shim,
    bool: shim,
    func: shim,
    number: shim,
    object: shim,
    string: shim,
    symbol: shim,

    any: shim,
    arrayOf: getShim,
    element: shim,
    elementType: shim,
    instanceOf: getShim,
    node: shim,
    objectOf: getShim,
    oneOf: getShim,
    oneOfType: getShim,
    shape: getShim,
    exact: getShim,

    checkPropTypes: emptyFunctionWithReset,
    resetWarningCache: emptyFunction
  };

  ReactPropTypes.PropTypes = ReactPropTypes;

  return ReactPropTypes;
};

var propTypes = createCommonjsModule(function (module) {
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

{
  // By explicitly using `prop-types` you are opting into new production behavior.
  // http://fb.me/prop-types-in-prod
  module.exports = factoryWithThrowingShims();
}
});

var controlers = function controlers(_ref) {
  var changeCategory = _ref.changeCategory,
      categories = _ref.categories;
  return /*#__PURE__*/React__default.createElement("div", {
    className: "filter-nav"
  }, /*#__PURE__*/React__default.createElement("button", {
    id: "btn",
    className: "btn btn-success active",
    onClick: changeCategory,
    "data-filter": "all"
  }, "All"), categories.map(function (category) {
    return /*#__PURE__*/React__default.createElement("button", {
      key: category,
      className: "btn btn-primary",
      onClick: changeCategory,
      "data-filter": category
    }, category);
  }));
};

controlers.propTypes = {
  categories: propTypes.arrayOf(propTypes.string),
  changeCategory: propTypes.func
};

function styleInject(css, ref) {
  if ( ref === void 0 ) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') { return; }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = ".filter .filter-gallery {\n  position: relative; }\n\n.filter .item {\n  display: block;\n  width: 200px;\n  margin: 5px;\n  z-index: 1; }\n  .filter .item img {\n    object-fit: cover;\n    width: 100%;\n    height: 100%; }\n\n.filter .item-panoramic {\n  width: 410px; }\n\n.filter .item-content {\n  position: relative;\n  width: 100%;\n  height: 100%; }\n  .filter .item-content img {\n    width: 100%;\n    height: auto; }\n\n@-webkit-keyframes closeWindow {\n  0% {\n    opacity: 1; }\n  100% {\n    opacity: 0; } }\n\n@keyframes closeWindow {\n  0% {\n    opacity: 1; }\n  100% {\n    opacity: 0; } }\n\n.ril__outer {\n  background-color: rgba(0, 0, 0, 0.85);\n  outline: none;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  z-index: 1000;\n  width: 100%;\n  height: 100%;\n  -ms-content-zooming: none;\n  -ms-user-select: none;\n  -ms-touch-select: none;\n  -ms-touch-action: none;\n  touch-action: none; }\n\n.ril__outerClosing {\n  opacity: 0; }\n\n.ril__inner {\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0; }\n\n.ril__image,\n.ril__imagePrev,\n.ril__imageNext {\n  position: absolute;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  margin: auto;\n  max-width: none;\n  -ms-content-zooming: none;\n  -ms-user-select: none;\n  -ms-touch-select: none;\n  -ms-touch-action: none;\n  touch-action: none; }\n\n.ril__imageDiscourager {\n  background-repeat: no-repeat;\n  background-position: center;\n  background-size: contain; }\n\n.ril__navButtons {\n  border: none;\n  position: absolute;\n  top: 0;\n  bottom: 0;\n  width: 20px;\n  height: 34px;\n  padding: 40px 30px;\n  margin: auto;\n  cursor: pointer;\n  opacity: 0.7; }\n\n.ril__navButtons:hover {\n  opacity: 1; }\n\n.ril__navButtons:active {\n  opacity: 0.7; }\n\n.ril__navButtonPrev {\n  left: 0;\n  background: rgba(0, 0, 0, 0.2) url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIgd2lkdGg9IjIwIiBoZWlnaHQ9IjM0Ij48cGF0aCBkPSJtIDE5LDMgLTIsLTIgLTE2LDE2IDE2LDE2IDEsLTEgLTE1LC0xNSAxNSwtMTUgeiIgZmlsbD0iI0ZGRiIvPjwvc3ZnPg==\") no-repeat center; }\n\n.ril__navButtonNext {\n  right: 0;\n  background: rgba(0, 0, 0, 0.2) url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIgd2lkdGg9IjIwIiBoZWlnaHQ9IjM0Ij48cGF0aCBkPSJtIDEsMyAyLC0yIDE2LDE2IC0xNiwxNiAtMSwtMSAxNSwtMTUgLTE1LC0xNSB6IiBmaWxsPSIjRkZGIi8+PC9zdmc+\") no-repeat center; }\n\n.ril__downloadBlocker {\n  position: absolute;\n  top: 0;\n  left: 0;\n  right: 0;\n  bottom: 0;\n  background-image: url(\"data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7\");\n  background-size: cover; }\n\n.ril__caption,\n.ril__toolbar {\n  background-color: rgba(0, 0, 0, 0.5);\n  position: absolute;\n  left: 0;\n  right: 0;\n  display: -webkit-box;\n  display: -ms-flexbox;\n  display: flex;\n  -webkit-box-pack: justify;\n  -ms-flex-pack: justify;\n  justify-content: space-between; }\n\n.ril__caption {\n  bottom: 0;\n  max-height: 150px;\n  overflow: auto; }\n\n.ril__captionContent {\n  padding: 10px 20px;\n  color: #fff; }\n\n.ril__toolbar {\n  top: 0;\n  height: 50px; }\n\n.ril__toolbarSide {\n  height: 50px;\n  margin: 0; }\n\n.ril__toolbarLeftSide {\n  padding-left: 20px;\n  padding-right: 0;\n  -webkit-box-flex: 0;\n  -ms-flex: 0 1 auto;\n  flex: 0 1 auto;\n  overflow: hidden;\n  text-overflow: ellipsis; }\n\n.ril__toolbarRightSide {\n  padding-left: 0;\n  padding-right: 20px;\n  -webkit-box-flex: 0;\n  -ms-flex: 0 0 auto;\n  flex: 0 0 auto; }\n\n.ril__toolbarItem {\n  display: inline-block;\n  line-height: 50px;\n  padding: 0;\n  color: #fff;\n  font-size: 120%;\n  max-width: 100%;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap; }\n\n.ril__toolbarItemChild {\n  vertical-align: middle; }\n\n.ril__builtinButton {\n  width: 40px;\n  height: 35px;\n  cursor: pointer;\n  border: none;\n  opacity: 0.7; }\n\n.ril__builtinButton:hover {\n  opacity: 1; }\n\n.ril__builtinButton:active {\n  outline: none; }\n\n.ril__builtinButtonDisabled {\n  cursor: default;\n  opacity: 0.5; }\n\n.ril__builtinButtonDisabled:hover {\n  opacity: 0.5; }\n\n.ril__closeButton {\n  background: url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMSIgd2lkdGg9IjIwIiBoZWlnaHQ9IjIwIj48cGF0aCBkPSJtIDEsMyAxLjI1LC0xLjI1IDcuNSw3LjUgNy41LC03LjUgMS4yNSwxLjI1IC03LjUsNy41IDcuNSw3LjUgLTEuMjUsMS4yNSAtNy41LC03LjUgLTcuNSw3LjUgLTEuMjUsLTEuMjUgNy41LC03LjUgLTcuNSwtNy41IHoiIGZpbGw9IiNGRkYiLz48L3N2Zz4=\") no-repeat center; }\n\n.ril__zoomInButton {\n  background: url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMCIgaGVpZ2h0PSIyMCI+PGcgc3Ryb2tlPSIjZmZmIiBzdHJva2Utd2lkdGg9IjIiIHN0cm9rZS1saW5lY2FwPSJyb3VuZCI+PHBhdGggZD0iTTEgMTlsNi02Ii8+PHBhdGggZD0iTTkgOGg2Ii8+PHBhdGggZD0iTTEyIDV2NiIvPjwvZz48Y2lyY2xlIGN4PSIxMiIgY3k9IjgiIHI9IjciIGZpbGw9Im5vbmUiIHN0cm9rZT0iI2ZmZiIgc3Ryb2tlLXdpZHRoPSIyIi8+PC9zdmc+\") no-repeat center; }\n\n.ril__zoomOutButton {\n  background: url(\"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMCIgaGVpZ2h0PSIyMCI+PGcgc3Ryb2tlPSIjZmZmIiBzdHJva2Utd2lkdGg9IjIiIHN0cm9rZS1saW5lY2FwPSJyb3VuZCI+PHBhdGggZD0iTTEgMTlsNi02Ii8+PHBhdGggZD0iTTkgOGg2Ii8+PC9nPjxjaXJjbGUgY3g9IjEyIiBjeT0iOCIgcj0iNyIgZmlsbD0ibm9uZSIgc3Ryb2tlPSIjZmZmIiBzdHJva2Utd2lkdGg9IjIiLz48L3N2Zz4=\") no-repeat center; }\n\n.ril__outerAnimating {\n  -webkit-animation-name: closeWindow;\n  animation-name: closeWindow; }\n\n@-webkit-keyframes pointFade {\n  0%,\n  19.999%,\n  100% {\n    opacity: 0; }\n  20% {\n    opacity: 1; } }\n\n@keyframes pointFade {\n  0%,\n  19.999%,\n  100% {\n    opacity: 0; }\n  20% {\n    opacity: 1; } }\n\n.ril__loadingCircle {\n  width: 60px;\n  height: 60px;\n  position: relative; }\n\n.ril__loadingCirclePoint {\n  width: 100%;\n  height: 100%;\n  position: absolute;\n  left: 0;\n  top: 0; }\n\n.ril__loadingCirclePoint::before {\n  content: '';\n  display: block;\n  margin: 0 auto;\n  width: 11%;\n  height: 30%;\n  background-color: #fff;\n  border-radius: 30%;\n  -webkit-animation: pointFade 800ms infinite ease-in-out both;\n  animation: pointFade 800ms infinite ease-in-out both; }\n\n.ril__loadingCirclePoint:nth-of-type(1) {\n  -webkit-transform: rotate(0deg);\n  transform: rotate(0deg); }\n\n.ril__loadingCirclePoint:nth-of-type(7) {\n  -webkit-transform: rotate(180deg);\n  transform: rotate(180deg); }\n\n.ril__loadingCirclePoint:nth-of-type(1)::before,\n.ril__loadingCirclePoint:nth-of-type(7)::before {\n  -webkit-animation-delay: -800ms;\n  animation-delay: -800ms; }\n\n.ril__loadingCirclePoint:nth-of-type(2) {\n  -webkit-transform: rotate(30deg);\n  transform: rotate(30deg); }\n\n.ril__loadingCirclePoint:nth-of-type(8) {\n  -webkit-transform: rotate(210deg);\n  transform: rotate(210deg); }\n\n.ril__loadingCirclePoint:nth-of-type(2)::before,\n.ril__loadingCirclePoint:nth-of-type(8)::before {\n  -webkit-animation-delay: -666ms;\n  animation-delay: -666ms; }\n\n.ril__loadingCirclePoint:nth-of-type(3) {\n  -webkit-transform: rotate(60deg);\n  transform: rotate(60deg); }\n\n.ril__loadingCirclePoint:nth-of-type(9) {\n  -webkit-transform: rotate(240deg);\n  transform: rotate(240deg); }\n\n.ril__loadingCirclePoint:nth-of-type(3)::before,\n.ril__loadingCirclePoint:nth-of-type(9)::before {\n  -webkit-animation-delay: -533ms;\n  animation-delay: -533ms; }\n\n.ril__loadingCirclePoint:nth-of-type(4) {\n  -webkit-transform: rotate(90deg);\n  transform: rotate(90deg); }\n\n.ril__loadingCirclePoint:nth-of-type(10) {\n  -webkit-transform: rotate(270deg);\n  transform: rotate(270deg); }\n\n.ril__loadingCirclePoint:nth-of-type(4)::before,\n.ril__loadingCirclePoint:nth-of-type(10)::before {\n  -webkit-animation-delay: -400ms;\n  animation-delay: -400ms; }\n\n.ril__loadingCirclePoint:nth-of-type(5) {\n  -webkit-transform: rotate(120deg);\n  transform: rotate(120deg); }\n\n.ril__loadingCirclePoint:nth-of-type(11) {\n  -webkit-transform: rotate(300deg);\n  transform: rotate(300deg); }\n\n.ril__loadingCirclePoint:nth-of-type(5)::before,\n.ril__loadingCirclePoint:nth-of-type(11)::before {\n  -webkit-animation-delay: -266ms;\n  animation-delay: -266ms; }\n\n.ril__loadingCirclePoint:nth-of-type(6) {\n  -webkit-transform: rotate(150deg);\n  transform: rotate(150deg); }\n\n.ril__loadingCirclePoint:nth-of-type(12) {\n  -webkit-transform: rotate(330deg);\n  transform: rotate(330deg); }\n\n.ril__loadingCirclePoint:nth-of-type(6)::before,\n.ril__loadingCirclePoint:nth-of-type(12)::before {\n  -webkit-animation-delay: -133ms;\n  animation-delay: -133ms; }\n\n.ril__loadingCirclePoint:nth-of-type(7) {\n  -webkit-transform: rotate(180deg);\n  transform: rotate(180deg); }\n\n.ril__loadingCirclePoint:nth-of-type(13) {\n  -webkit-transform: rotate(360deg);\n  transform: rotate(360deg); }\n\n.ril__loadingCirclePoint:nth-of-type(7)::before,\n.ril__loadingCirclePoint:nth-of-type(13)::before {\n  -webkit-animation-delay: 0ms;\n  animation-delay: 0ms; }\n\n.ril__loadingContainer {\n  position: absolute;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0; }\n\n.ril__imagePrev .ril__loadingContainer,\n.ril__imageNext .ril__loadingContainer {\n  display: none; }\n\n.ril__errorContainer {\n  position: absolute;\n  top: 0;\n  right: 0;\n  bottom: 0;\n  left: 0;\n  display: -webkit-box;\n  display: -ms-flexbox;\n  display: flex;\n  -webkit-box-align: center;\n  -ms-flex-align: center;\n  align-items: center;\n  -webkit-box-pack: center;\n  -ms-flex-pack: center;\n  justify-content: center;\n  color: #fff; }\n\n.ril__imagePrev .ril__errorContainer,\n.ril__imageNext .ril__errorContainer {\n  display: none; }\n\n.ril__loadingContainer__icon {\n  color: #fff;\n  position: absolute;\n  top: 50%;\n  left: 50%;\n  -webkit-transform: translateX(-50%) translateY(-50%);\n  transform: translateX(-50%) translateY(-50%); }\n";
styleInject(css);

var filterContainer = /*#__PURE__*/function (_Component) {
  _inherits(filterContainer, _Component);

  var _super = _createSuper(filterContainer);

  function filterContainer() {
    var _this;

    _classCallCheck(this, filterContainer);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "state", {
      filterBy: 'all',
      width: _this.onResize,
      isOpen: false,
      photoIndex: 0,
      id: uuid(),
      imgSrcArray: []
    });

    _defineProperty(_assertThisInitialized(_this), "calculateLayout", function () {
      var _this$props = _this.props,
          columnWidth = _this$props.columnWidth,
          gutterHeight = _this$props.gutterHeight,
          gutterWidth = _this$props.gutterWidth,
          responsive = _this$props.responsive,
          itemHeight = _this$props.itemHeight,
          lightbox = _this$props.lightbox;
      var _this$state = _this.state,
          filterBy = _this$state.filterBy,
          width = _this$state.width;
      var childrenNodesArray = React__default.Children.toArray(_this.props.children);
      var imgSrcArray = [];
      var columns = responsive ? Math.min(Math.floor(width / (columnWidth + gutterWidth)), _this.props.columns) : _this.props.columns;
      var filteredChildQty = 0;
      if (!_this.validateChildrenNodes(childrenNodesArray)) return;
      var childrenDataArray = childrenNodesArray.map(function (childrenNode) {
        if (filterBy === 'all' || childrenNode.props["data-filter"] === filterBy) {
          var column = filteredChildQty % columns;
          var row = Math.floor(filteredChildQty / columns);
          var x = column * columnWidth + column * gutterWidth;
          var y = row * itemHeight + row * gutterHeight;
          filteredChildQty++;
          lightbox && imgSrcArray.push(childrenNode.props.src);
          return {
            x: x,
            y: y,
            children: childrenNode
          };
        }
      });
      var gridHeight = Math.ceil(filteredChildQty / columns) * (itemHeight + gutterHeight) - gutterHeight || 0;
      return _this.renderTheLayout({
        childrenDataArray: childrenDataArray,
        gridHeight: gridHeight,
        imgSrcArray: imgSrcArray
      });
    });

    _defineProperty(_assertThisInitialized(_this), "renderTheLayout", function (_ref) {
      var childrenDataArray = _ref.childrenDataArray,
          gridHeight = _ref.gridHeight,
          imgSrcArray = _ref.imgSrcArray;
      var _this$props2 = _this.props,
          columnWidth = _this$props2.columnWidth,
          itemHeight = _this$props2.itemHeight,
          duration = _this$props2.duration,
          lightbox = _this$props2.lightbox,
          easing = _this$props2.easing;
      var id = _this.state.id;
      var photoIdex = 0;
      return /*#__PURE__*/React__default.createElement("div", {
        className: "filter-gallery",
        style: {
          height: gridHeight
        }
      }, /*#__PURE__*/React__default.createElement(reactTransitionGroup.TransitionGroup, {
        component: null
      }, childrenDataArray.map(function (childData, index) {
        if (childData === undefined) return;
        photoIdex++;
        var innerPhotoIndex = photoIdex;
        return /*#__PURE__*/React__default.createElement(reactTransitionGroup.Transition, {
          key: index + id,
          "in": true,
          timeout: {
            enter: duration,
            exit: 0
          }
        }, function (state) {
          return /*#__PURE__*/React__default.createElement("div", {
            className: "item",
            style: {
              transition: "all ".concat(duration, "ms ").concat(easing),
              position: 'absolute',
              width: "".concat(columnWidth, "px"),
              height: "".concat(itemHeight, "px"),
              top: childData.y ? childData.y : undefined,
              left: childData.x ? childData.x : undefined
            },
            onClick: lightbox ? _this.openLightboxHandler(imgSrcArray, innerPhotoIndex) : undefined
          }, childData.children);
        });
      })));
    });

    _defineProperty(_assertThisInitialized(_this), "changeCategory", function (e) {
      _this.setState({
        filterBy: e.target.dataset.filter
      });
    });

    _defineProperty(_assertThisInitialized(_this), "validateChildrenNodes", function (childrenNodesArray) {
      var lightbox = _this.props.lightbox;
      var valid = true;
      childrenNodesArray.forEach(function (children) {
        if (!children.props["data-filter"]) {
          throw new Error('Expected children to have data-filter attribute');
          valid = false;
        }

        if (!children.props.src && lightbox) {
          throw new Error('Expected children to have src attribute');
          valid = false;
        }
      });
      return valid;
    });

    _defineProperty(_assertThisInitialized(_this), "openLightboxHandler", function (imgSrcArray, index) {
      return function () {
        _this.setState({
          isOpen: true,
          photoIndex: index - 1,
          imgSrcArray: imgSrcArray
        });
      };
    });

    _defineProperty(_assertThisInitialized(_this), "onResize", function (width) {
      _this.setState({
        width: width
      });
    });

    _defineProperty(_assertThisInitialized(_this), "lightboxCloseHandler", function () {
      _this.setState({
        isOpen: false
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onMovePrevRequest", function () {
      var _this$state2 = _this.state,
          imgSrcArray = _this$state2.imgSrcArray,
          photoIndex = _this$state2.photoIndex;

      _this.setState({
        photoIndex: (photoIndex + imgSrcArray.length - 1) % imgSrcArray.length
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onMoveNextRequest", function () {
      var _this$state3 = _this.state,
          imgSrcArray = _this$state3.imgSrcArray,
          photoIndex = _this$state3.photoIndex;

      _this.setState({
        photoIndex: (photoIndex + 1) % imgSrcArray.length
      });
    });

    return _this;
  }

  _createClass(filterContainer, [{
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          categories = _this$props3.categories,
          lightbox = _this$props3.lightbox;
      var _this$state4 = this.state,
          isOpen = _this$state4.isOpen,
          imgSrcArray = _this$state4.imgSrcArray,
          photoIndex = _this$state4.photoIndex;
      return /*#__PURE__*/React__default.createElement(ReactResizeDetector, {
        handleWidth: true,
        onResize: this.onResize
      }, /*#__PURE__*/React__default.createElement("div", {
        className: "filter filter-basic"
      }, /*#__PURE__*/React__default.createElement(controlers, {
        categories: categories,
        changeCategory: this.changeCategory
      }), this.calculateLayout(), lightbox && isOpen && /*#__PURE__*/React__default.createElement(Lightbox, {
        mainSrc: imgSrcArray[photoIndex],
        nextSrc: imgSrcArray[(photoIndex + 1) % imgSrcArray.length],
        prevSrc: imgSrcArray[(photoIndex + imgSrcArray.length - 1) % imgSrcArray.length],
        onCloseRequest: this.lightboxCloseHandler,
        onMovePrevRequest: this.onMovePrevRequest,
        onMoveNextRequest: this.onMoveNextRequest
      })));
    }
  }]);

  return filterContainer;
}(React.Component);

filterContainer.defaultProps = {
  categories: [],
  gutterWidth: 0,
  gutterHeight: 0,
  layout: 'simple',
  component: 'div',
  itemHeight: 150,
  columnWidth: 200,
  columns: 5,
  duration: 300,
  responsive: true,
  lightbox: false,
  easing: "ease-out"
};
filterContainer.propTypes = {
  categories: propTypes.arrayOf(propTypes.string),
  columns: propTypes.number.isRequired,
  columnWidth: propTypes.number.isRequired,
  itemHeight: propTypes.number,
  gutterWidth: propTypes.number,
  gutterHeight: propTypes.number,
  duration: propTypes.number.isRequired,
  easing: propTypes.string,
  component: propTypes.string,
  responsive: propTypes.bool.isRequired,
  lightbox: propTypes.bool.isRequired
};

module.exports = filterContainer;
