// MDB-React-Calendar/src/plugin/components/BaseTableCell.js
import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

const BaseTableCell = ({
  children,
  day,
  double,
  event,
  head,
  notCurrMonth,
  today,
  list
}) => {
  function runEvent() {
    if (head) return;

    event({}, day);
  }

  const classes = classNames(
    "table-cell",
    list && "table-cell--list",
    head && "table-cell__head",
    today && "table-cell--today",
    notCurrMonth && "table-cell--not-curr-month",
    double && "table-cell--double"
  );

  const Tag = list ? "li" : "div";

  return (
    <Tag
      tabIndex={!head ? 0 : -1}
      className={classes}
      data-day={day}
      onClick={runEvent}
    >
      {children}
    </Tag>
  );
};

BaseTableCell.propTypes = {
  children: PropTypes.node,
  day: PropTypes.instanceOf(Date),
  double: PropTypes.bool,
  event: PropTypes.func,
  head: PropTypes.bool,
  notCurrMonth: PropTypes.bool,
  today: PropTypes.bool,
  list: PropTypes.bool
};

export default BaseTableCell;
