// MDB-React-Color-Picker/src/components/controllers.js
import React, { PureComponent } from 'react'
import { convert2rgba } from '../utils/'
import PropTypes from "prop-types";

class Controllers extends PureComponent {
    onInputHandler = e => {
        const inputValue = e.target.value;
        const rgbaColor = convert2rgba(inputValue);

        this.props.saveColor(rgbaColor, inputValue);
    }

    copyToClipBoard = () => {
        const el = document.createElement('textarea');
        el.value = this.props.currentColorLabel;
        document.body.appendChild(el);
        el.select();
        document.execCommand('copy');
        document.body.removeChild(el);
    }

    save = () => {
        this.props.saveColor()
        this.copyToClipBoard();
    }

    clear = () => {
        this.props.clearColor()
    }

    render() {
        return (
            <div className="mdb-cp-interaction">
                <input className="mdb-cp-result mb-1" type="text" spellCheck="false" value={this.props.currentColorLabel} onChange={this.onInputHandler} />

                <input className="mdb-cp-type mb-1" data-active={this.props.colorSpace === "hex"} value="HEX" type="button" onClick={this.props.changeColorSpace('hex')} />
                <input className="mdb-cp-type mb-1" data-active={this.props.colorSpace === "rgba"} value="RGBa" type="button" onClick={this.props.changeColorSpace('rgba')} />
                <input className="mdb-cp-type mb-1" data-active={this.props.colorSpace === "hsla"} value="HSLa" type="button" onClick={this.props.changeColorSpace('hsla')} />
                <input className="mdb-cp-type mb-1" data-active={this.props.colorSpace === "hsva"} value="HSVa" type="button" onClick={this.props.changeColorSpace('hsva')} />
                <input className="mdb-cp-type mb-1" data-active={this.props.colorSpace === "cmyk"} value="CMYK" type="button" onClick={this.props.changeColorSpace('cmyk')} />


                <input className="mdb-cp-save mb-1" value={this.props.saveLabel} onClick={this.save} type="button" />
                <input className="mdb-cp-clear mb-1" value={this.props.clearLabel} type="button" onClick={this.clear} />
            </div >

        )
    }
}

Controllers.propTypes = {
    currentColorLabel: PropTypes.string.isRequired,
    saveColor: PropTypes.func.isRequired,
    clearColor: PropTypes.func.isRequired,
    changeColorSpace: PropTypes.func.isRequired,
    saveLabel: PropTypes.string.isRequired,
    clearLabel: PropTypes.string.isRequired,
    colorSpace: PropTypes.string.isRequired,
}


export default Controllers
