// MDB-React-Table-Editor/src/plugin/components/DeleteModal.js
import React from "react";
import PropTypes from "prop-types";
import {
  MDBBtn,
  MDBModal,
  MDBModalBody,
  MDBModalHeader,
  MDBModalFooter
} from "mdbreact";

const DeleteModal = ({ isOpen, toggle, action, label, desc, yes, no }) => (
  <MDBModal isOpen={isOpen} toggle={toggle}>
    <MDBModalHeader className="flex-center">
      <span className="font-weight-bold text-center text-danger">{label}</span>
    </MDBModalHeader>
    <MDBModalBody>
      <p className="text-center h5">{desc}</p>
    </MDBModalBody>
    <MDBModalFooter className="flex-center">
      <MDBBtn color="danger" onClick={action}>
        {yes}
      </MDBBtn>
      <MDBBtn color="primary" onClick={toggle}>
        {no}
      </MDBBtn>
    </MDBModalFooter>
  </MDBModal>
);

DeleteModal.propTypes = {
  label: PropTypes.string.isRequired,
  desc: PropTypes.string.isRequired,
  yes: PropTypes.string.isRequired,
  no: PropTypes.string.isRequired,
  action: PropTypes.func.isRequired,
  isOpen: PropTypes.bool.isRequired,
  toggle: PropTypes.func.isRequired
};

export default DeleteModal;
