// MDB-React-Table-Editor/src/plugin/components/Modal.js
import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  MDBBtn,
  MDBModal,
  MDBModalBody,
  MDBModalHeader,
  MDBModalFooter,
  MDBInput
} from "mdbreact";

class Modal extends Component {
  constructor(props) {
    super(props);

    const row = this.props.editor
      ? this.props.rows.find(row => row.id === this.props.checkedRows[0])
      : {};

    this.state = { ...row };
  }

  inputChange = e => this.setState({ [e.target.name]: e.target.value });

  submitForm = e => {
    e.preventDefault();
  };

  actionSubmit = () => {
    this.props.action({ ...this.state });
    this.props.toggle();
  };

  render() {
    const { columns, okLabel, cancelLabel, isOpen, toggle } = this.props;

    return (
      <MDBModal isOpen={isOpen} toggle={toggle}>
        <MDBModalHeader toggle={toggle}>
          <span className="font-weight-bold text-secondary">{okLabel}</span>
        </MDBModalHeader>
        <MDBModalBody>
          <form onSubmit={this.submitForm}>
            {columns.map(col => (
              <MDBInput
                key={col.field}
                label={col.label}
                name={col.field}
                value={this.state[col.field]}
                onInput={this.inputChange}
              />
            ))}
          </form>
        </MDBModalBody>
        <MDBModalFooter>
          <MDBBtn color="secondary" onClick={toggle}>
            {cancelLabel}
          </MDBBtn>
          <MDBBtn color="primary" onClick={this.actionSubmit}>
            {okLabel}
          </MDBBtn>
        </MDBModalFooter>
      </MDBModal>
    );
  }
}

Modal.propTypes = {
  cancelLabel: PropTypes.string.isRequired,
  okLabel: PropTypes.string.isRequired,
  action: PropTypes.func.isRequired,
  checkedRows: PropTypes.arrayOf(PropTypes.string).isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  editor: PropTypes.bool.isRequired,
  isOpen: PropTypes.bool.isRequired,
  rows: PropTypes.arrayOf(PropTypes.object).isRequired,
  toggle: PropTypes.func.isRequired
};

export default Modal;
